### PR TITLE
🌱 Make release notes tool not dependent on local git

### PR DESCRIPTION
**Pull request by g-gaston**
**What this PR does / why we need it**:

Now all the data is retrieved through GitHub APIs, making the tool more portable and easier to use. It should not increase the rate limiting chances since it now performs less API requests (by getting the label from all PRs at once instead of with one request per PR).

This tool had also outgrown its previous code structure, which made a bit difficult to made these changes and think about how the overall program works. So I went a bit overboard with the original purpose and refactor the code a bit. Mostly splitting it into different files and adding some comments here and there. the new structure should make easier to reuse piece of the code from other release team tools and compose the different pieces to do similar but different tasks (like the weekly update).

I also have simplified the command flags for the tool, since the comms team doesn't need most of them and they needed to be re-implemented with this new "API approach" if we wanted to keep supporting them. We can implement support and re-add them if the need for them comes back in the future.

The new UI is:
```
Usage of ./bin/notes:
  -add-kubernetes-version-support
        If enabled, will add the Kubernetes version support header. (default true)
  -branch string
        The branch to generate the notes from. If not set, it will be calculated from release.
  -deprecation
        If enabled, will add a templated deprecation warning header. (default true)
  -from string
        The tag or commit to start from. It must be formatted as heads/<branch name> for branches and tags/<tag name> for tags. If not set, it will be calculated from release.
  -pre-release-version
        If enabled, will add a pre-release warning header. (default false)
  -prefix-area-label
        If enabled, will prefix the area label. (default true)
  -release string
        The tag for the enw release.
  -repository string
        The repo to run the tool from. (default "kubernetes-sigs/cluster-api")
  -to string
        The ref (tag, branch or commit to stop at. It must be formatted as heads/<branch name> for branches and tags/<tag name> for tags. If not set, it will default to branch.
```

For new releases, it's possible to just specify the tag of the new (to be released) release and the program will compute the rest of the values. This also opens the door in the future to alter the behavior depending on if it's a RC/beta release (instead of having individual flags).

```sh
./bin/notes --release=v1.6.0 > CHANGELOG/v1.6.0.md
```

Apologies in advance for the big diff.

I want to merge #9617 first so we can look at the output changes.

/area release
cc @kubernetes-sigs/cluster-api-release-team

### CODE REVIEW

File: hack/tools/release/notes/test/golden/v1.5.0.md
Comment (g-gaston) [inline hack/tools/release/notes/test/golden/v1.5.0.md:62]: These the same PR. The title was changed in the PR and since now we read from the PR API and not the commits, now we output the new title. This is expected (and IMO a good change).

File: hack/tools/release/notes/main.go
Comment (willie-yao) [inline hack/tools/release/notes/main.go:63]: nit: It might be helpful to specify the format of the repo

File: hack/tools/release/notes/main.go
Comment (willie-yao) [inline hack/tools/release/notes/main.go:67]:
```suggestion
	flag.StringVar(&config.newTag, "release", "", "The tag for the new release.")
```

File: hack/tools/release/notes/print.go
Comment (willie-yao) [inline hack/tools/release/notes/print.go:148]: Shouldn't this block also work if `len(mergeslice) == 1`? I could be misunderstanding how `strings.Join()` works in Go though

**Review by willie-yao — COMMENTED** [review]
This is some great stuff! Excited to see this tool in use. Everything looks good pending a few minor comments.

File: hack/tools/release/notes/github.go
Comment (g-gaston) [reply hack/tools/release/notes/github.go:56]: In this case this is the max page for this api
there is no need to change dynamically the page size. Also, a smaller page size would increase the chances of rate limiting, not decrease it

File: hack/tools/release/notes/main.go
Comment (g-gaston) [reply hack/tools/release/notes/main.go:63]: isn't the default enough? or what are you looking for?

File: hack/tools/release/notes/print.go
Comment (g-gaston) [reply hack/tools/release/notes/print.go:148]: The message changes
`there has been 1 contribution` vs `there has been 1 contributions`

File: hack/tools/release/notes/main.go
Comment (willie-yao) [reply hack/tools/release/notes/main.go:63]: Ah that's my bad. I forgot the default will show up with -h

**Comment by furkatgofurov7** [comment]
@g-gaston looks like it needs a rebase, I'll take a look after it is rebased.

**Comment by sbueringer** [comment]
(didn't look at the code yet) 

@g-gaston Does this also make it possible to use tide with the merge method so folks don't have to manually squash commits before PR merge? (xref: https://github.com/kubernetes-sigs/cluster-api/issues/9249)

**Comment by g-gaston** [comment]
> (didn't look at the code yet)
> 
> @g-gaston Does this also make it possible to use tide with the merge method so folks don't have to manually squash commits before PR merge? (xref: #9249)

maybe. At least it should make it easier to make it work if it doesn't on its current state
@mjlshen was gonna verify that but I'll take a look when I find some time

File: hack/tools/release/notes/test/golden/v1.5.0.md
Comment (g-gaston) [inline hack/tools/release/notes/test/golden/v1.5.0.md:56]: This is bc the the bug fix from https://github.com/kubernetes-sigs/cluster-api/pull/9814

File: hack/tools/release/notes/test/golden/v1.5.0.md
Comment (g-gaston) [inline hack/tools/release/notes/test/golden/v1.5.0.md:20]: The 4 additions here is bc we now include PRs that use tide squashing :)

File: hack/tools/release/notes/github.go
Comment (sbueringer) [inline hack/tools/release/notes/github.go:88]: copy&paste error

File: hack/tools/release/notes/list.go
Comment (sbueringer) [inline hack/tools/release/notes/list.go:88]: Does this work for PRs merged into release branches? (as we're hard-coding main as the base branch)

File: hack/tools/release/notes/print.go
Comment (sbueringer) [inline hack/tools/release/notes/print.go:40]: I think there are words missing in this comment

**Review by sbueringer — COMMENTED** [review]
Probably fine overall. To be honest, I don't have the bandwidth to properly review this entirely on top of other work

File: hack/tools/release/notes/main.go
Comment (g-gaston) [reply hack/tools/release/notes/main.go:36]: It's not. This seems like an outdated comment, I'll change it.
Btw, nothing should have changed in that regard. As long as you configure the `--branch` and `--from` `--to` appropriately (based on your usecase), the command will return the same PRs as before regardless if the next tag exists or not.

File: hack/tools/release/notes/github.go
Comment (g-gaston) [reply hack/tools/release/notes/github.go:1]: I decided to keep using `gh` for now since that was the previous dependency so everyone who was using it already had it setup. It already handles authentication so I didn't have to document/add config how to authenticate against github.

I think switching this to a github library is a good improvement, it should simplify this code. However, I would prefer to have it as an improvement on top of this, to minimize disruption by this first refactor.

File: hack/tools/release/notes/list.go
Comment (g-gaston) [reply hack/tools/release/notes/list.go:88]: it does work yeah, that's what `l.branch` is (it lists PRs both to main and to `l.branch`)

We include `main` here because when cutting a new minor version, there will PRs merged in both main and the release branch. The returned PRs from this list are filtered from the list of commits in between the refs `to` and `from`, so it won't print PRs from `main` that belong to a different release, even if merged during the same date range. 

This is just an optimization to avoid listing PRs over all branches, since we know we don't need the PRs from other release branches.

File: hack/tools/release/notes/main.go
Comment (sbueringer) [reply hack/tools/release/notes/main.go]: Alright. Sounds good, thank you!

File: hack/tools/release/notes/github.go
Comment (sbueringer) [reply hack/tools/release/notes/github.go:1]: Sounds good.

File: hack/tools/release/notes/list.go
Comment (sbueringer) [reply hack/tools/release/notes/list.go:88]: Sounds good. Can you add a godoc comment explaingin this? Thank you :)

**Comment by sbueringer** [comment]
Looks good, thank you! One remaining comment: https://github.com/kubernetes-sigs/cluster-api/pull/9618#discussion_r1423966218

I guess we don't need the squash anymore :)

File: hack/tools/release/notes/list.go
Comment (g-gaston) [reply hack/tools/release/notes/list.go:99]: Done :)

**Review by willie-yao — APPROVED** [review]
/lgtm

**Comment by sbueringer** [comment]
@g-gaston Thank you, looks good!

Really appreciate the improvements made to the tool in this PR (and over the last few months)!!

Happy to approve after a rebase for the go.mod conflicts.

**Comment by g-gaston** [comment]
> @g-gaston Thank you, looks good!
> 
> Really appreciate the improvements made to the tool in this PR (and over the last few months)!!
> 
> Happy to approve after a rebase for the go.mod conflicts.

Done!

**Comment by sbueringer** [comment]
@g-gaston Could it be that we have to update release-tasks.md? As far as I can tell the current instructions don't work anymore without modifications

File: hack/tools/release/notes/main.go
Comment (sbueringer) [inline hack/tools/release/notes/main.go:156]: I tried to run this tool for CR with the following args:
```
-from tags/v0.16.0 -branch main -release v0.17.0 -repository kubernetes-sigs/controller-runtime
```

This func returns in l.158. Then l.182 is never executed and it eventually panics in parseRef

```go
panic: runtime error: index out of range [1] with length 1

goroutine 1 [running]:
main.parseRef({0x0, 0x0})
        /Users/buringerst/code/src/sigs.k8s.io/cluster-api/hack/tools/release/notes/ref.go:43 +0xbc
main.(*notesCmd).run(0x1400004e050)
        /Users/buringerst/code/src/sigs.k8s.io/cluster-api/hack/tools/release/notes/main.go:101 +0x124
main.main()
        /Users/buringerst/code/src/sigs.k8s.io/cluster-api/hack/tools/release/notes/main.go:41 +0x28
Exiting.


```

Comment (sbueringer) [reply hack/tools/release/notes/main.go:156]: Otherwise it works great!

Comment (g-gaston) [reply hack/tools/release/notes/main.go:156]: You are 100% right
I'll get it fixed

That said, you don't need to specify `-from` in this case. Specifying `-release` is enough for the program to infer the rest. You do need to specify the branch bc it seems like in `controller-runtime` minor releases are cut directly from `main`. That would work even without the fix.

```sh
./bin/notes --branch main --release v0.17.0 --repository kubernetes-sigs/controller-runtime --prefix-area-label=false --add-kubernetes-version-support=false
```

Comment (sbueringer) [reply hack/tools/release/notes/main.go:156]: Thx, yup sounds reasonable. I just started with adding --from without thinking much about it and then I added what was needed :)

**Comment by g-gaston** [comment]
> @g-gaston Could it be that we have to update release-tasks.md? As far as I can tell the current instructions don't work anymore without modifications

yes totally
https://github.com/kubernetes-sigs/cluster-api/pull/9957

File: hack/tools/release/notes/main.go
Comment (g-gaston) [reply hack/tools/release/notes/main.go:156]: https://github.com/kubernetes-sigs/cluster-api/pull/9958

**Comment by cahillsf** [comment]
hey @g-gaston 👋 just checking, is [this ](https://github.com/kubernetes-sigs/cluster-api/issues/9249) still needed or has your PR accomplished this?

**Comment by g-gaston** [comment]
> hey @g-gaston 👋 just checking, is [this ](https://github.com/kubernetes-sigs/cluster-api/issues/9249) still needed or has your PR accomplished this?

This PR accomplishes it :)

**Comment by sbueringer** [comment]
Thx again for that!!

**Pull request merged**
